src/Features/CSharp/Portable/ReplaceMethodWithProperty/CSharpReplaceMethodWithPropertyService.cs

```diff
@@ -0,0 +1,272 @@
+﻿using System;
+using System.Composition;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.ReplaceMethodWithProperty;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ReplaceMethodWithProperty
+{
+    [ExportLanguageService(typeof(IReplaceMethodWithPropertyService), LanguageNames.CSharp), Shared]
+    internal class CSharpReplaceMethodWithPropertyService : IReplaceMethodWithPropertyService
+    {
+        public string GetMethodName(SyntaxNode methodNode)
+        {
+            return ((MethodDeclarationSyntax)methodNode).Identifier.ValueText;
+        }
+
+        public SyntaxNode GetMethodDeclaration(SyntaxToken token)
+        {
+            var containingMethod = token.Parent.FirstAncestorOrSelf<MethodDeclarationSyntax>();
+            if (containingMethod == null)
+            {
+                return null;
+            }
+
+            var start = containingMethod.AttributeLists.Count > 0
+                ? containingMethod.AttributeLists.Last().GetLastToken().GetNextToken().SpanStart
+                : containingMethod.SpanStart;
+
+            // Offer this refactoring anywhere in the signature of the method.
+            var position = token.SpanStart;
+            if (position < start || position > containingMethod.ParameterList.Span.End)
```

Comment:
Test boundary cases:
Invoke from first token that's acceptable and the last token that's admissible (caret at left and right of close paren token of method declaration).